### PR TITLE
🚧 [claude-inline-subtasks] deepseek: live subagent tiles and children [step 4/5]

### PR DESCRIPTION
**Pull request by AlexDochioiu**
## Complexity

🚧 Complex — live lifecycle, cross-turn correlation, child catalogs, and a coordinated managed-adapter upgrade.

## What

- Project correlated starts/ends into one direct-parent child-linked tile with authoritative launch prompt and bounded terminal state; root activity rolls up through descendants.
- Track delegation correlation until both standard-tool and child lifecycles finish; consolidate deferred call/update handling and retain generic startup failures.
- Merge persisted/live children through the session service, retaining root-project ownership.
- Require protocol v2; remove the temporary v1 set, mapper version flag, gating branches, and obsolete fallback test. Keep live-state reset at connection initialization.
- Pin the managed target and minimum accepted runtime to **0.1.3**, using all six verified published archive digests. Reject outdated explicit binaries and replace outdated PATH candidates with the managed runtime.

**Size: 1,556 changed lines**, including tests, documentation, and the user-requested release pinning. This is approximately the 1,500-line target. Step 4/5 replacing #1293, following #1298, #1301, and #1304.

## Why

The user requested publishing/pinning the Sesori-owned adapter now rather than retaining compatibility solely for release sequencing. [Adapter release-prep #16](https://github.com/sesori-ai/sesori-deepseek-acp/pull/16) merged and [v0.1.3 is published](https://github.com/sesori-ai/sesori-deepseek-acp/releases/tag/v0.1.3). Replay remains step 5; scoped interruption remains a later feature PR.

## Risk and test focus

- Highest risk: cross-turn/terminal correlation, duplicate cards, child attribution, and runtime installation/version enforcement.
- No database/schema change, replay projection, or interrupt consumption in this slice.
- No dual-version adapter compatibility remains. Standard ACP remains version 1; the DeepSeek extension requires version 2.
- Package release verification is complete; the plan's final feature/client E2E matrix is still pending and is not claimed here.

## Expected result

The explicit Install action downloads adapter 0.1.3. Startup only selects an already-present compatible PATH runtime or the exact managed 0.1.3 runtime; it does not automatically download upgrades. Users with only an older managed adapter must use Install before DeepSeek can start. An outdated explicit binary must be updated or its override removed. The consumer requires v2 initialization and shows live child-linked tiles. No database impact.

## Verification

- DeepSeek `dart analyze --fatal-infos`: clean.
- DeepSeek `dart test`: **81 passed**, including 0.1.2 explicit rejection/PATH replacement and v1/v3 initialize rejection.
- Original live implementation architecture/correctness reviews: no findings.
- Formatting and `git diff --check`: passed.
- [Release run 33960157438](https://github.com/sesori-ai/sesori-deepseek-acp/actions/runs/33960157438): all six native package builds/smokes, consumer conformance, release-set verification, and publish succeeded.
- Independently downloaded all six assets, checked their SHA-256 against the published manifest and pinned values, inspected archive paths/link safety and 0.1.3 package/build metadata. Build source `3976bcd1afbb5072bec3e87cb12e9f603282bcb6`; conformance consumer `69e9478d681d97d6a0f1b2d7b1f0baf03a0cd71c`.
- Both protocol fixture versions, DTO serializers, and history repository untouched.

### CODE REVIEW

**Comment by AlexDochioiu** [comment]
[Sesori reply] Codex cloud review is unavailable due to its usage limit; this is not a code finding. Audited the six lint suppressions: four cover heterogeneous standard ACP update/parameter maps in the single deferred snapshot; two cover reading open initialize metadata before typed API parsing. Each is line-scoped and justified in place. DeepSeek analyze and all 82 tests pass; architecture and focused correctness reviews found no issues. CI and the remaining review channel are pending.

File: bridge/sesori_plugin_deepseek/lib/src/deepseek_event_mapper.dart [hunk 1]
```diff
@@ -39,8 +160,149 @@ class DeepSeekEventMapper({
...
+    if (byToolCall?.isEmpty ?? false) _deferredDelegations.remove(sessionId);
+    return deferred;
+  }
+
+  static bool _isDelegationTitle({required String title}) => title == "subagent" || title == "subagent_fork";
+
+  static bool _isTerminalToolStatus({required String status}) => status == "completed" || status == "failed";
+
+  List<BridgeSseEvent> _mapSubagentEnded({required DeepSeekSubagentEndedDto notification}) {
+    delegationTracker.markChildEnded(childSessionId: notification.childSessionId);
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply bridge/sesori_plugin_deepseek/lib/src/deepseek_event_mapper.dart:243]: [Sesori reply] Declining without a reachable producer flow. In the authoritative native adapter, sessions.ts #childStarted records the direct parent, and #childEnded emits sessionId from that same record.parentId (not from caller-supplied end data). Both travel on the owned ordered output path. A wrong-parent frame requires a separate adapter corruption bug; this is not an independently authorized client request. The malformed-completion recovery intentionally finalizes the named known child rather than leaving it busy. Adding another correlation guard for invented identity corruption is outside the proportionality rule for this slice; no production change.

File: bridge/sesori_plugin_deepseek/lib/src/deepseek_event_mapper.dart [hunk 2]
```diff
@@ -39,8 +160,149 @@ class DeepSeekEventMapper({
...
+    } on Object catch (error, stackTrace) {
+      final recoveryEvents = _recoverMalformedSubagent(notification: notification);
+      Log.w("[deepseek] ignored malformed sub-agent notification", error, stackTrace);
+      return recoveryEvents;
+    }
+  }
+
+  List<BridgeSseEvent> _mapSubagentStarted({required DeepSeekSubagentStartedDto notification}) {
+    _removeDeferredDelegation(sessionId: notification.sessionId, toolCallId: notification.toolCallId);
+    final events = mapChildSpawned(
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply bridge/sesori_plugin_deepseek/lib/src/deepseek_event_mapper.dart:179]: [Sesori reply] Declining the proposed terminal-failure tombstone. Native #childStarted runs within the executing delegation scope and queues started on the parent's outputTail; the standard tool result uses that same serialized tail through #projectEvent/#projectChildEvent. A startup failure has no child start, while a successful start is queued before its terminal result. The update-before-call handling does not establish that a terminal result can overtake started on this transport. No current sequence produces the proposed late start, so retaining an additional lifetime registry would add speculative state.

File: bridge/sesori_plugin_deepseek/lib/src/repositories/trackers/deepseek_delegation_tracker.dart [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  }
+
+  void forgetSession({required String sessionId}) {
+    final childDelegation = _byChild[sessionId];
+    if (childDelegation != null) _remove(delegation: childDelegation);
+    final parentDelegations = _byParent.remove(sessionId);
+    if (parentDelegations == null) return;
+    for (final delegation in parentDelegations.values) {
+      _byChild.remove(delegation.childSessionId);
+    }
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply bridge/sesori_plugin_deepseek/lib/src/repositories/trackers/deepseek_delegation_tracker.dart:70]: [Sesori reply] The real deletion owner already handles every descendant. AcpPlugin clears the named session and loops childSessionTracker.childSessionIds(sessionId: sessionId), calling eventMapper.forgetSession(childId) for each descendant before dropping tracker records (acp_plugin.dart:1806-1814). childSessionIds explicitly returns running and finished descendants, not just direct children. Therefore nested DeepSeek correlations are removed through that existing loop, including grandchildren whose parent has finished. Repeating a recursive traversal in the delegation index would duplicate ownership and cannot fix an uncovered production path here.

File: docs/regression/tools-and-file-changes.md [hunk 4]
```diff
@@ -49,7 +49,13 @@ signal that a tool changed files.
...
   identity, bounded presenter output, terminal result/error state, and diff
   content. Presenter failure degrades to a generic bounded tool card instead of
-  dropping the call or result.
+  dropping the call or result. With protocol v2, correlated sub-agent starts
+  replace exact `subagent`/`subagent_fork` cards with one child-linked tile;
+  identifiable updates arriving before their call are deferred too. Start/end
+  events retain the direct parent's transcript and keep root activity busy.
+  Launch prompts remain authoritative without child prompt echoes; malformed
+  ends finalize known children as error. Startup failures retain one generic
+  terminal card. Protocol v1 keeps generic cards; v2 replay projection is pending.
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply docs/regression/tools-and-file-changes.md]: [Sesori reply] Fixed in f04bb3eee. The paragraph now explicitly says the v2 live path is implemented and unit-tested but unreleased, with adapter 0.1.3 publication and the final E2E matrix pending. This does not claim passing release/E2E coverage; capability cells and runtime target/floor remain unchanged.

File: bridge/sesori_plugin_deepseek/lib/src/deepseek_event_mapper.dart [hunk 5]
```diff
@@ -39,8 +160,149 @@ class DeepSeekEventMapper({
...
+      };
+    } on Object catch (error, stackTrace) {
+      final recoveryEvents = _recoverMalformedSubagent(notification: notification);
+      Log.w("[deepseek] ignored malformed sub-agent notification", error, stackTrace);
+      return recoveryEvents;
+    }
+  }
+
+  List<BridgeSseEvent> _mapSubagentStarted({required DeepSeekSubagentStartedDto notification}) {
+    _removeDeferredDelegation(sessionId: notification.sessionId, toolCallId: notification.toolCallId);
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply bridge/sesori_plugin_deepseek/lib/src/deepseek_event_mapper.dart:178]: [Sesori reply] Checked the concrete rejection paths: API parsing rejects invalid/empty IDs before this method, and _mapSubagent rejects deleted parent/child IDs before invoking it. AcpChildSessionTracker.spawn only returns null for those deleted IDs; mapChildSpawned's other early return is the already-validated empty-ID case. These calls execute synchronously, with no yield between the checks and spawn. Thus no current accepted DTO can lose its deferred card through a rejected spawn here. Malformed starts instead follow _recoverMalformedSubagent and flush the retained card. Declining a change based on a hypothetical future rejection branch.

File: bridge/sesori_plugin_deepseek/lib/src/deepseek_event_mapper.dart [hunk 6]
```diff
@@ -21,8 +25,125 @@ class DeepSeekEventMapper({
...
+      final String value => value,
+      _ => null,
+    };
+    final status = switch (update["status"]) {
+      final String value => value,
+      _ => null,
+    };
+    final hasDelegationTitle = title != null && _isDelegationTitle(title: title);
+    switch (update["sessionUpdate"]) {
+      case "tool_call" when hasDelegationTitle:
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply bridge/sesori_plugin_deepseek/lib/src/deepseek_event_mapper.dart:111]: [Sesori reply] Checked the actual producer rather than adding another defensive case. The native adapter's projectSessionEvent tool/call branch initializes title from event.data.name (sessions.ts:795-802), and the pinned delegation tools supply subagent/subagent_fork, not a missing-title initial frame. Partial tool_call_update frames are the title-optional path, and they already consult delegationStarted. A titleless initial tool_call is not emitted by the supported adapter and is not a demonstrated reorder flow. No guard added for that fabricated packet; an observed different delegation presenter/wire shape would warrant revisiting the producer boundary.

File: docs/regression/questions-and-permissions.md [hunk 7]
```diff
@@ -41,8 +41,9 @@ reaches the backend so the turn continues.
   option, it cancels rather than escalating or selecting another scope. The
   pinned CLI does not forward `ask_user` over ACP, so Copilot declares no
   question capability and Sesori never invents a custom question channel.
-- DeepSeek standard ACP permissions use the request's explicit session ID when
-  present and retain the ACP active-turn fallback when an agent omits it. They
+- DeepSeek standard ACP permissions use explicit session IDs or exact tracked
+  tool-call attribution, retained across parent turns. Ambiguity cancels; only
+  unattributed requests use the ACP active-turn fallback. They
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply docs/regression/questions-and-permissions.md]: [Sesori reply] Fixed in f04bb3eee: the sentence now explicitly says These permissions preserve the exact tool call ID, removing the ambiguous pronoun.

**Comment by AlexDochioiu** [comment]
[Sesori reply] Inspected all seven findings and replied on each thread. f04bb3eee fixes the unreleased-v2 coverage wording and ambiguous permission subject. Declined five code changes after tracing native parent identity/output ordering, the existing recursive deletion owner, synchronous spawn rejection checks, and the actual delegation title producer; none identified a reachable missing behavior. Production code is unchanged from the 82-test/analyzer passing head. Documentation diff checks pass; total size is now 1,475 changed lines. Awaiting current-head CI/review evidence; adapter publication, replay, and final E2E gates remain unchanged.

**Comment by AlexDochioiu** [comment]
[Sesori reply] User-directed release change: 69e9478d6 removes temporary v1 compatibility; the consumer now requires v2. Analyzer and 81 tests pass. The six lint suppressions are unchanged apart from line numbers and remain justified at the heterogeneous ACP boundary. This PR is deliberately draft and NOT ready to merge: native release preparation now precedes this merge, and real 0.1.3 asset hashes plus target/minimum enforcement must land here after publication. The old runtime manifest is retained only in this intermediate draft head; replay stays step 5.

**Comment by AlexDochioiu** [comment]
[Sesori reply] Release pinning is complete: v0.1.3 is published, all six package builds/smokes and release gates passed, and every downloaded archive SHA-256/build metadata was verified against the pinned assets. This revision sets managed target and accepted minimum to 0.1.3. Tests prove explicit 0.1.2 rejection and PATH replacement; protocol-v1 compatibility is gone. Analyzer and all 81 tests pass. The PR is out of draft and awaits current-head CI/review. Final feature E2E and replay/scoped-stop follow-ups remain separate.

**Comment by AlexDochioiu** [comment]
[Sesori reply] Correction to the rollout description: ManagedRuntimeProvisionService only resolves already-present runtimes. Downloads happen through the explicit Install command and ManagedRuntimeInstallService, not automatically on startup. Users with only an older managed version must invoke Install for 0.1.3; an outdated explicit override must be updated or removed. The expected-result section now states this accurately. The PATH fallback test proves selection of an already-present managed 0.1.3, not automatic downloading. Automatic upgrades would be a separate behavior change.

File: bridge/sesori_plugin_deepseek/lib/src/deepseek_event_mapper.dart [hunk 8]
```diff
@@ -21,8 +25,121 @@ class DeepSeekEventMapper({
...
+      case "tool_call_update"
+          when delegationStarted ||
+              hasDelegationTitle ||
+              (_deferredDelegations[sessionId]?.containsKey(toolCallId) ?? false):
+        if (delegationStarted) {
+          if (status != null && _isTerminalToolStatus(status: status)) {
+            delegationTracker.markToolTerminal(parentSessionId: sessionId, toolCallId: toolCallId);
+          }
+          return const [];
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply bridge/sesori_plugin_deepseek/lib/src/deepseek_event_mapper.dart:120]: [Sesori reply] Checked the pinned native producer and the shared mapper boundary. DeepSeek projectSessionEvent emits an explicit assistantMessageId(sessionId, turn, step) on text and reasoning chunks. Subsequent model steps have distinct message IDs, which AcpEventMapper._chunkIdentity preserves. The normal tool boundary cited here only closes/finalizes the synthetic id-less envelope (_finalizeCurrentIdlessAssistantText / _closeCurrentIdlessAssistantContent); DeepSeek does not use that envelope for these chunks. Thus bypassing that id-less boundary cannot concatenate the preceding and following native assistant steps. No boundary hook or synthetic renumbering added without a real supported-producer trace that demonstrates the issue.

File: bridge/sesori_plugin_deepseek/lib/src/deepseek_event_mapper.dart [hunk 9]
```diff
@@ -21,8 +25,121 @@ class DeepSeekEventMapper({
...
+      final String value => value,
+      _ => null,
+    };
+    final status = switch (update["status"]) {
+      final String value => value,
+      _ => null,
+    };
+    final hasDelegationTitle = title != null && _isDelegationTitle(title: title);
+    switch (update["sessionUpdate"]) {
+      case "tool_call" when hasDelegationTitle:
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply bridge/sesori_plugin_deepseek/lib/src/deepseek_event_mapper.dart:111]: [Sesori reply] This repeats the earlier titleless-initial-call finding without a new producer flow. The pinned adapter initializes tool_call.title from event.data.name, and its delegation tools emit subagent/subagent_fork. Partial tool_call_update frames are the title-optional path and already consult delegationStarted. We handle identifiable reordered updates, not fabricated initial packets from a different producer. Retaining the previous reasoned decline; no additional defensive condition.

File: bridge/sesori_plugin_deepseek/lib/src/deepseek_event_mapper.dart [hunk 10]
```diff
@@ -21,8 +25,121 @@ class DeepSeekEventMapper({
...
+  final Map<String, Map<String, _DeferredDeepSeekDelegation>> _deferredDelegations = {};
+
+  void resetLiveState() {
+    _deferredDelegations.clear();
+    delegationTracker.clear();
+  }
+
+  @override
+  void beginTurn({required String sessionId, required String? messageId}) {
+    _deferredDelegations.remove(sessionId);
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply bridge/sesori_plugin_deepseek/lib/src/deepseek_event_mapper.dart:37]: [Sesori reply] The two lifetimes are deliberate: accepted child correlation survives parent turns in DeepSeekDelegationTracker, while pre-start generic snapshots are turn-local. The native session/prompt response awaits terminal admission, agent.whenIdle(), and record.outputTail in #settle before completing; the shared ACP plugin serializes prompts for that session. A normal startup failure therefore emits its terminal tool update and flushes the generic card before the next beginTurn. A background child may continue across turns, but it already has an accepted start and uses the separate persistent tracker. No current flow was identified that needs pre-start snapshots carried into an unrelated next turn.

File: .plan/active/claude-inline-subtasks/TRACKER.md [hunk 11]
```diff
@@ -650,3 +650,18 @@ check lives in its DTO factory, consistent with replay metadata; semantic checks
 remain in the API. Architecture implementation review approved with no findings.
 Initialization remains v1-only until slice 4 installs the live consumer. No
 fixture bytes, runtime pins, interrupt consumer, or live/replay projection changed.
+
+Slice 4 verification: architecture and focused correctness reviews found no
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply .plan/active/claude-inline-subtasks/TRACKER.md:654]: [Sesori reply] Added in 887fd26cb: the tracker now pins the exact 747fbd3eb9..97adc68c8 range and records +1,446/-100 = 1,546 lines, with a reproducible numstat command. It explicitly explains the 46-line soft-budget overage as the user-requested runtime publication/pinning in this PR. Later review bookkeeping is labeled additional rather than making the historical number self-referential.

**Comment by AlexDochioiu** [comment]
[Sesori reply] 887fd26cb merges actual origin/main without rebasing. The capability-table conflict preserves both upstream Codex tiles and this PR’s DeepSeek support. CI’s failure was three tests indexing the newly typed PluginSessionStatus as JSON; they now assert the sealed values. DeepSeek analyzer and all 81 tests pass after the merge. Inspected all four latest review threads: added reproducible historical size evidence and declined the three producer-incompatible lifecycle/boundary cases with concrete code references. Automatic runtime upgrades remain outside this PR, with the separate session. Awaiting current-head CI/review.
